### PR TITLE
PR to fix issue #1269 - add plain text parser to core

### PR DESCRIPTION
**Pull request by lfurzewaddock**
PR to fix issue #1269 - add plain text parser to core

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/custom-parser.test.js [hunk 1]
```diff
@@ -8,6 +8,29 @@ const sget = require('simple-get').concat
...
+    setImmediate(parse, body)
+  }
+  function parse (json) {
+    try {
+      callback(null, body)
+    } catch (err) {
+      callback(err, null)
+    }
+  }
+}
```
Comment (mcollina) [inline test/custom-parser.test.js]: can you add space here?

File: test/custom-parser.test.js [hunk 2]
```diff
@@ -8,6 +8,29 @@ const sget = require('simple-get').concat
 const Fastify = require('..')
 
 const jsonParser = require('fast-json-body')
+// Adapted from https://github.com/delvedor/fast-json-body/blob/master/json-body.js
```
Comment (mcollina) [inline test/custom-parser.test.js]: can you add a space before the comment?

File: test/custom-parser.test.js [hunk 3]
```diff
@@ -8,6 +8,29 @@ const sget = require('simple-get').concat
...
+function plainTextParser (request, callback) {
+  var body = ''
+  request.on('error', onError)
```
Comment (mcollina) [inline test/custom-parser.test.js]: for completeness, you need to call `request.setEncoding('utf8')`. I'm not sure why this is not failing.

File: lib/contentTypeParser.js [hunk 4]
```diff
@@ -191,6 +195,16 @@ function defaultJsonParser (req, body, done) {
   done(null, json)
 }
 
+function defaultPlainTextParser (req, body, done) {
+  try {
+    var plainText = body
+  } catch (err) {
+    err.statusCode = 400
+    return done(err, undefined)
+  }
```
Comment (mcollina) [inline lib/contentTypeParser.js]: I don't think this could _ever_ throw, so the try-catch should be removed.

Comment (cemremengu) [reply lib/contentTypeParser.js]: Should this be `toString` instead?

```js
var plainText = body.toString()
```

File: test/custom-parser.test.js (hunk 2 above)
Comment (cemremengu) [reply test/custom-parser.test.js]: I dont think comment is necessary (imo)

File: lib/contentTypeParser.js (hunk 4 above)
Comment (lfurzewaddock) [reply lib/contentTypeParser.js]: I thought about using `.toString()` too @cemremengu, but decided to follow the defaultJsonParser function example. However, I could check for null and return a new error based upon `FST_ERR_CTP_EMPTY_JSON_BODY()` or just use `.toString()`?

File: test/custom-parser.test.js (hunk 3 above)
Comment (lfurzewaddock) [reply test/custom-parser.test.js]: `request.setEncoding('utf8')` is not present in @delvedor source, but I'll add it.

File: test/custom-parser.test.js (hunk 2 above)
Comment (lfurzewaddock) [reply test/custom-parser.test.js]: @mcollina @cemremengu should I remove the source credit comment?

Comment (cemremengu) [reply test/custom-parser.test.js]: Since @delvedor is a core mainter of fastify I dont think it will be a problem for him :)

Comment (lfurzewaddock) [reply test/custom-parser.test.js]: ;-)

**Comment by lfurzewaddock** [comment]
Hi @mcollina, @cemremengu. I've pushed the changes.

File: test/custom-parser.test.js [hunk 5]
```diff
@@ -8,6 +8,31 @@ const sget = require('simple-get').concat
...
+  }
+  function onEnd () {
+    setImmediate(parse, body)
+  }
+  function parse (json) {
+    try {
+      callback(null, body)
+    } catch (err) {
+      callback(err, null)
+    }
```
Comment (mcollina) [inline test/custom-parser.test.js]: this try-catch is useless, can you please remove?

File: test/custom-parser.test.js (hunk 1 above)
Comment (lfurzewaddock) [reply test/custom-parser.test.js]: done.

File: test/custom-parser.test.js (hunk 2 above)
Comment (lfurzewaddock) [reply test/custom-parser.test.js]: done and comment removed.

File: test/custom-parser.test.js (hunk 3 above)
Comment (lfurzewaddock) [reply test/custom-parser.test.js]: done.

File: test/custom-parser.test.js [hunk 6]
```diff
@@ -8,6 +8,27 @@ const sget = require('simple-get').concat
...
+  request.on('data', onData)
+  request.on('end', onEnd)
+  function onError (err) {
+    callback(err, null)
+  }
+  function onData (chunk) {
+    body += chunk
+  }
+  function onEnd () {
+    setImmediate(parse, body)
```
Comment (mcollina) [inline test/custom-parser.test.js]: you can just do `callback(null, body)` no need to use `parse()` or `setImmediate()`

Comment (lfurzewaddock) [reply test/custom-parser.test.js]: OK, with code changed to...

```
function plainTextParser (request, callback) {
  var body = ''
  request.setEncoding('utf8')
  request.on('error', onError)
  request.on('data', onData)
  function onError (err) {
    callback(err, null)
  }
  function onData (chunk) {
    body += chunk
  }
  callback(null, body)
}
```

...a unit test fails;
```
test/custom-parser.test.js ........................ 130/131
  Can override the default plain text parser
  not ok should be equal
    +++ found
    --- wanted
    -hello world
    compare: ===
    at:
      line: 584
      column: 9
      file: test/custom-parser.test.js
      function: sget
    stack: |
      sget (test/custom-parser.test.js:584:9)
      concat (node_modules/simple-get/index.js:89:7)
      IncomingMessage.<anonymous> (node_modules/simple-concat/index.js:7:13)
    source: |
      t.strictEqual(body.toString(), 'hello world')

```

Comment (mcollina) [reply test/custom-parser.test.js]: you need to maintain onEnd()

Comment (lfurzewaddock) [reply test/custom-parser.test.js]: cheers @mcollina 
done. change pushed.

Comment (cemremengu) [reply test/custom-parser.test.js]: @lfurzewaddock Flaky CI. Fixed now.

Comment (lfurzewaddock) [reply test/custom-parser.test.js]: :-)

**Comment by cemremengu** [comment]
@lfurzewaddock I think there should be a test for verifying that the body is string when there is no custom parser defined and content-type is `plain/text` (i.e. verify that the default parser works as expected)

**Comment by lfurzewaddock** [comment]
@cemremengu, @mcollina I've written 4 tests attempting to break the `defaultPlainTextParser` function but cannot. 

For every test the `body` variable within `defaultPlainTextParser` function is always a string.

File: test/custom-parser.test.js (hunk 6 above)
Comment (lfurzewaddock) [reply test/custom-parser.test.js]: done.

File: test/custom-parser.test.js [hunk 7]
```diff
@@ -673,6 +673,113 @@ test('Should get the body as string', t => {
...
+    sget({
+      method: 'POST',
+      url: 'http://localhost:' + fastify.server.address().port,
+      headers: {
+        'Content-Type': 'text/plain'
+      }
+    }, (err, response, body) => {
+      t.error(err)
+      t.strictEqual(response.statusCode, 200)
+      t.strictEqual(typeof (body.toString()), 'string')
```
Comment (cemremengu) [inline test/custom-parser.test.js]: On top off my head, `typeof (body.toString())` does not make sense 🤔  

it should be `typeof body` only probably?

Comment (lfurzewaddock) [reply test/custom-parser.test.js]: Yes, that was a bit of a dead end as the body at that point was always a buffer object.

I'll remove .toString() and test for type object.

Comment (cemremengu) [reply test/custom-parser.test.js]: @lfurzewaddock maybe I am wrong though, lets wait for @mcollina

Comment (lfurzewaddock) [reply test/custom-parser.test.js]: done.

**Pull request merged**